### PR TITLE
Add porting guide and documentation for changes to argument spec validation

### LINKED ISSUE #73983: argument spec refactoring breaks some things
### Summary

I'm seeing some breakage in collections caused (very likely) by #73703:
1. `AttributeError: 'AnsibleModule' object has no attribute '_check_type_dict'` (community.general, infoblox.nios_modules);
2. `ImportError: cannot import name 'handle_aliases'` (from `ansible.module_utils.common.parameters`) (community.crypto, community.sops).

While 1. could be considered 'own fault' because it is a private API, 2. looks like a bug, since that's a break in a public interface.

### Issue Type

Bug Report

### Component Name

core

### Ansible Version

devel

### Configuration

.

### OS / Environment

.

### Steps to Reproduce

.

### Expected Results

.

### Actual Results

.

### PR DESCRIPTION
**Pull request by samdoran**
##### SUMMARY

Add detailed information about breaking changes to the 2.11 porting guide. Update the `module_utils` doc to include documentation for a lot more of the public interfaces.

Fixes #73983.

##### ISSUE TYPE

- Docs Pull Request

##### COMPONENT NAME

`docs/docsite/rst/porting_guides/porting_guide_core_2.11.rst`
`docs/docsite/rst/reference_appendices/module_utils.rst`
`lib/ansible/module_utils/common/arg_spec.py`
`lib/ansible/module_utils/common/parameters.py`

### CODE REVIEW

**Comment by samdoran** [comment]
Our docs sanity test does not like specifying optional parameters.

```rst

:param: some_param
:type some_param: list, optional
```

I'm pretty sure that is valid in Sphinx.


```
/root/ansible/lib/ansible/module_utils/common/arg_spec.py:docstring of ansible.module_utils.common.arg_spec.ArgumentSpecValidator:: WARNING: py:class reference target not found: optional
```

**Comment by abadger** [comment]
> Our docs sanity test does not like specifying optional parameters.
> 
> ```
> :param: some_param
> :type some_param: list, optional
> ```
> 
> I'm pretty sure that is valid in Sphinx.
> 
> ```
> /root/ansible/lib/ansible/module_utils/common/arg_spec.py:docstring of ansible.module_utils.common.arg_spec.ArgumentSpecValidator:: WARNING: py:class reference target not found: optional
> ```
> 
> It renders correctly, but the sanity test doesn't like it.

As a workaround, you can use:
``` rst
:arg positional_manadatory_arg: Description of argument
:type: str
:kwarg keyword_optional_arg: Description of argument
:type list:
```

I've never used the `:param:` syntax before so I don't know what cornercase this might be running into.

**Comment by abadger** [comment]
Another thing to try would be to document the class construction parameters in the `__init__()` method's docstring rather than in the class docstring.  It might be that using `:type: foo, bar` in the `:py:class:` directive isn't working right but it would work right in the `:py:meth:` and `:py:func:` directives.

**Comment by samdoran** [comment]
> I've never used the :param: syntax before so I don't know what cornercase this might be running into.

My understanding is that `arg`, `kwarg`, and `param` all mean the same thing and render the same in the output. That's what I gather from [the field lists documentation](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists). My testing shows that they all get grouped under **Parameters** in the rendered doc.

> Another thing to try would be to document the class construction parameters in the __init__() method's docstring rather than in the class docstring. 

That didn't change anything.

I think the root of this is that specifying `optional` is incorrect. I found that example [here](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html#the-sphinx-docstring-format) and maybe it is wrong.

File: lib/ansible/module_utils/common/arg_spec.py
Comment (webknjaz) [inline lib/ansible/module_utils/common/arg_spec.py]: Try
```suggestion
        :type mutually_exclusive: Optional[list]
```

**Comment by webknjaz** [comment]
@samdoran @abadger Sphinx build runs on py3-only already right? Can we add https://pypi.org/project/sphinx-autodoc-typehints/ and use type hints instead of putting types in docstrings?

File: lib/ansible/module_utils/common/arg_spec.py
Comment (webknjaz) [inline lib/ansible/module_utils/common/arg_spec.py]: maybe render it shorter?
```suggestion
        sanitized with :func:`~ansible.module_utils.common.parameters.sanitize_keys` before logging or displaying.
```

File: lib/ansible/module_utils/errors.py
Comment (webknjaz) [inline lib/ansible/module_utils/errors.py]: pls only put a title on the first line per PEP257

File: lib/ansible/module_utils/common/arg_spec.py
Comment (webknjaz) [inline lib/ansible/module_utils/common/arg_spec.py]: Sounds like it's
```suggestion
        :type required_by: dict[str, list[str]]
```

File: lib/ansible/module_utils/common/arg_spec.py
Comment (webknjaz) [inline lib/ansible/module_utils/common/arg_spec.py]: is this
```suggestion
        :type required_one_of: list[list[str]]
```

File: lib/ansible/module_utils/common/arg_spec.py
Comment (webknjaz) [inline lib/ansible/module_utils/common/arg_spec.py]: Or maybe
```suggestion
        :type required_together: list or None
```
per https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#python-signatures

File: lib/ansible/module_utils/common/arg_spec.py
Comment (samdoran) [reply lib/ansible/module_utils/common/arg_spec.py]: Wouldn't this require having imported `Optional` from `typing`? Using this fails the docs-build sanity test.

**Comment by abadger** [comment]
> > I've never used the :param: syntax before so I don't know what cornercase this might be running into.
> 
> My understanding is that `arg`, `kwarg`, and `param` all mean the same thing and render the same in the output. That's what I gather from [the field lists documentation](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists). My testing shows that they all get grouped under **Parameters** in the rendered doc.
> 
<nod> The reason I suggested using kwarg is that in that style, you don't have to say `, optional`.  I suppose you don't have to change from param as the keyword and can just leave optional off. (the rendered result is the same.  The wording "Default value is" in the description and the definition which shows there is a default value will both show that it is optional.

**Comment by abadger** [comment]
Hmmmm.... One other thing that I thing isn't clear and is being hidden in the discussion of using type hints....

`optional` is not the same as `typing.Optional`.

When I read:
```
:param mutually_exclusive: Set to a list of pairs of parameters which are mutually exclusive with each other
:type: sequence of tuples, optional
```
I take it to mean that mutually_exclusive doesn't need to be specified.

`t.Optional` actually means what @webknjaz said `list of tuples or None`.

Delving into the implementation a bit (even though, that isn't what this PR is for ;-)   I think it would have been better to default to an empty tuple rather than None.  The two reasons I think that's better API is that (1) you then only accept sequences here, not sequences or None.  This is a bit less for API consumers to remember. and (2) It simplifies the logic of the API implementors.  If it's always a sequence, your code to process it looks like this:
``` python
for pairs in mutually_exclusive:
    # do thing
```

If you use None as the default value, you instead have to code like this:
``` python
if mutually_exclusive:
    for pairs in mutually_exclusive:
        # do thing
```

**Comment by samdoran** [comment]
> I take it to mean that mutually_exclusive doesn't need to be specified.

Yup, that's what I was trying to convey. At this point, I'm just going to leave it to the function signature to convey that since we don't currently have a clear way to express it.

> I think it would have been better to default to an empty tuple rather than None.

That's a very good point! I tend to default to `None` or a string to avoid the [mutable defaults problem](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments)). I definitely should have used a tuple. 🤦‍♂️

**Comment by webknjaz** [comment]
@abadger yep, you are right. I actually often just say that it's optional in text. Also, the defaults are normally visible in the function signature line and it is pretty obvious what that means...

**Comment by webknjaz** [comment]
Interestingly https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists has an example with `priority=1` in the signature and `int or Note` in the docstring 🤯

File: lib/ansible/module_utils/common/arg_spec.py
Comment (webknjaz) [inline lib/ansible/module_utils/common/arg_spec.py]: it's probably this?
```suggestion
        :type mutually_exclusive: list[str] or list[list[str]]
```

File: docs/docsite/rst/reference_appendices/module_utils.rst
Comment (webknjaz) [inline docs/docsite/rst/reference_appendices/module_utils.rst:38]: Side note: It's possible to add `.. currentmodule:: ansible.module_utils` at the top and avoid prefixing it in every definition.

**Comment by webknjaz** [comment]
@samdoran if you'll go the type hints way, make sure you include the imports. Example: https://github.com/ansible/ansible/blob/3740d7b/lib/ansible/galaxy/dependency_resolution/providers.py#L11-L21.

File: docs/docsite/rst/reference_appendices/module_utils.rst
Comment (samccann) [inline docs/docsite/rst/reference_appendices/module_utils.rst:47]: Since we aren't at 3.2 or above, is this going to cause display problems for 2.11 docs?

**Comment by samccann** [comment]
Also, there seems to be a lot here that impacts collection maintainers. Consider adding an entry into the next Ansible Bullhorn if you haven't already.

**Review by samccann — APPROVED** [review]
docs portion LGTM

File: docs/docsite/rst/reference_appendices/module_utils.rst
Comment (samdoran) [reply docs/docsite/rst/reference_appendices/module_utils.rst:47]: No, it will display just fine with Sphinx 2.1.2, but only because no other private things have docstrings.

This comment is mainly to document the behavior that took me quite some time to figure out what was going on. `:private-members:` was rendering the docs for _all_ private methods/attributes even if they weren't in the list. It's because it wasn't until Sphinx 3.2 that `:private-members:` honored the values passed in.

**Comment by samdoran** [comment]
> Also, there seems to be a lot here that impacts collection maintainers. Consider adding an entry into the next Ansible Bullhorn if you haven't already.

I opened issues on every collection that was using the private methods mentioned in the porting guide.

**Pull request merged**
